File: src/main/java/com/hiberus/anaya/redmineapi/Issue.java
package com.hiberus.anaya.redmineapi;

import org.json.JSONObject;

import java.io.IOException;

/**
 * A redmine issue
 */
public final class Issue {

    /* ------------------------- manager ------------------------- */

    private final RedmineManager manager; // an issue is associated to a manager

    /* ------------------------- data ------------------------- */

    /**
     * The identifier
     */
    public final int id;
    /**
     * The name of its project
     */
    public final String project;
    /**
     * The subject (title)
     */
    public final String subject;
    /**
     * The description, may probably be empty
     */
    public final String description;

    private double estimated_hours; // estimated hours
    private int done_ratio; // realization percentage
    private double spentHours; // total hours, uninitialized

    private final JSONObject original; // the original raw object, for diff purposes

    /* ------------------------- constructors ------------------------- */

    Issue(JSONObject rawIssue, RedmineManager manager) {
        // parse issue from raw json data
        this.manager = manager;
        original = rawIssue;

        id = rawIssue.getInt("id");
        project = rawIssue.getJSONObject("project").optString("name", "");
        subject = rawIssue.optString("subject", "");
        description = rawIssue.optString("description");
        estimated_hours = rawIssue.optDouble("estimated_hours", RedmineManager.NONE);
        done_ratio = rawIssue.optInt("done_ratio", 0);
        spentHours = rawIssue.optDouble("spent_hours", RedmineManager.UNINITIALIZED);
    }

    /* ------------------------- properties getters ------------------------- */

    /**
     * Get total hours spent on this issue.
     * Can be not initialized
     */
    public double getSpent() {
        return spentHours;
    }

    /**
     * Estimated hours
     */
    public double getEstimated() {
        return estimated_hours;
    }

    /**
     * Ratio of realization
     */
    public int getRealization() {
        return done_ratio;
    }

    /**
     * @return the url to see this issue details
     */
    public String getUrl() {
        return manager.domain + "issues/" + id;
    }

    /**
     * @return a short string describing this issue
     * @see #toString()
     */
    public String toShortString() {
        return "#" + id + ": " + subject;
    }

    /**
     * @return a fairly long string describing this issue
     * @see #toShortString()
     */
    @Override
    public String toString() {
        return project + "\n" + toShortString();
    }

    /* ------------------------- properties setters ------------------------- */

    /**
     * Changes the total hours spent on this issue
     *
     * @param amount number oh hours (negative to substract)
     */
    public void addSpent(double amount) {
        assert spentHours >= 0;
        spentHours += amount;
        assert spentHours >= 0;
    }

    public void addEstimated(double amount) {
        // still uninitialized, cancel
        if (estimated_hours == RedmineManager.UNINITIALIZED) return;

        if (estimated_hours == 0 && amount < 0) {
            // no hours and want to subtract, disable
            estimated_hours = RedmineManager.NONE;
            return;
        }
        if (estimated_hours == RedmineManager.NONE) {
            // disabled and want to change, set to 0 if it wants to add, keep if not
            if (amount > 0) estimated_hours = 0;
            return;
        }

        // else, change
        amount = Math.max(estimated_hours + amount, 0) - estimated_hours; // don't subtract what can't be subtracted
        estimated_hours += amount;
    }

    public void addRealization(int amount) {
        amount = clamp(0, done_ratio + amount, 100) - done_ratio; // don't subtract what can't be subtracted, and don't add what can be added
        done_ratio += amount;
    }

    public void syncRealization() {
        done_ratio = clamp(0, (int) (spentHours / estimated_hours * 100), 100);
    }

    /* ------------------------- downloading ------------------------- */

    /**
     * Loads spent hours (nothing if already initialized)
     * Long operation
     *
     * @throws IOException on network error
     */
    public void downloadSpent() throws IOException {
        if (spentHours == RedmineManager.UNINITIALIZED) {
            spentHours = UrlJSONKt.get(manager.domain + "issues/" + id + ".json?key=" + manager.key)
                    .getJSONObject("issue").optDouble("spent_hours", RedmineManager.NONE);
        }
    }

    /* ------------------------- uploading ------------------------- */

    /**
     * @return a list of changes made to this issue, as an object (empty means no changes)
     */
    public JSONObject getChanges() {
        JSONObject changes = new JSONObject();

        // TODO: use a JSON as container so this can be a simple foreach diff
        if (original == null || estimated_hours != original.optDouble("estimated_hours", RedmineManager.NONE)) {
            // changed hours
            changes.put("estimated_hours", estimated_hours == RedmineManager.NONE ? "" : estimated_hours);
        }
        if (original == null || done_ratio != original.getDouble("done_ratio")) {
            // changed comment
            changes.put("done_ratio", done_ratio);
        }

        // return changes (empty for nothing)
        return changes;
    }

    /**
     * @return true if this entry requires upload, false otherwise
     */
    public boolean requiresUpload() {
        return !(
                getChanges().isEmpty() // no changes, no upload
        );
    }

    /**
     * Uploads an entry to redmine (unless not needed)
     *
     * @throws IOException on upload error
     */
    public void uploadTimeEntry() throws IOException {
        // get changes
        JSONObject changes = getChanges();

        // ignore unmodified
        if (changes.isEmpty()) return;

        // update
        System.out.println("Updating issue " + id + " with data: " + changes);
        if (RedmineManager.OFFLINE) return;
        if (UrlJSONKt.put(manager.domain + "issues/" + id + ".json?key=" + manager.key, new JSONObject().put("issue", changes)) != 200) {
            throw new IOException("Error when updating issue " + id + " with data: " + changes);
        }
    }

    /* ------------------------- utils ------------------------- */

    private static int clamp(int min, int value, int max) {
        return Math.min(Math.max(value, min), max);
    }
}

File: src/main/kotlin/com/hiberus/anaya/redmineapi/Issue.kt
package com.hiberus.anaya.redmineapi

import org.json.JSONObject
import java.io.IOException

/** A redmine issue */
class Issue {

    /* ------------------------- manager ------------------------- */

    private val remote: Remote // for remote interactions

    /* ------------------------- data ------------------------- */

    /** The identifier */
    val id: Int

    /** The name of its project */
    val project: String
    private val projectId: Int // TODO: move to a Project class

    /** The subject (title) */
    val subject: String

    /** The description, may probably be empty */
    val description: String

    /**
     * Estimated hours, greater than 0
     * null if unset
     */
    var estimated: Double? = null
        set(value) {
            field = value?.coerceAtLeast(0.0)
        }

    /** Ratio of realization (percentage) */
    var realization: Int
        private set

    /**
     * Total hours spent on this issue.
     * null if uninitialized
     */
    val spent get() = remote_spent?.plus(local_spent)
    private var remote_spent: Double? // the remote part, may be uninitialized
    private var local_spent = 0.0 // the local part, must be the sum of all entries

    /**
     * [spent]/[estimated]
     * null if invalid
     */
    val spent_realization: Int?
        get() {
            return ((spent?.takeIf { it >= 0 } ?: return null) / (estimated?.takeIf { it > 0 } ?: return null) * 100).toInt()
        }

    /** ID of the assigned user */
    val assigned_to: Int?

    /** Journal entries (non-empty) */
    var journals: List<String> = emptyList()

    /** the original raw object, if any, for diff purposes */
    private val original: JSONObject?

    /* ------------------------- constructors ------------------------- */

    @Suppress("ConvertSecondaryConstructorToPrimary")
    internal constructor(rawIssue: JSONObject, remote: Remote) {
        this.remote = remote
        original = rawIssue
        id = rawIssue.getInt("id")
        project = rawIssue.getJSONObject("project").optString("name")
        projectId = rawIssue.getJSONObject("project").optInt("id")
        subject = rawIssue.optString("subject", "")
        description = rawIssue.optString("description")
        estimated = rawIssue.noNaNDouble("estimated_hours")
        realization = rawIssue.optInt("done_ratio", 0)
        remote_spent = rawIssue.noNaNDouble("spent_hours")
        assigned_to = rawIssue.optJSONObject("assigned_to")?.getInt("id")
    }

    /* ------------------------- properties ------------------------- */

    /** the url to see this issue details */
    val url get() = remote.getIssueDetailsUrl(this)

    /**
     * @return a short string describing this issue
     * @see .toString
     */
    fun toShortString() = "#$id: $subject${remote.userId?.takeIf { it == assigned_to }?.let { " [you]" } ?: ""}"

    /**
     * @return a multiline string describing this issue
     * @see .toShortString
     */
    override fun toString() = "$project\n#$id: $subject" + (remote.userId?.takeIf { it == assigned_to }?.let { "\nAssigned to you" } ?: "")

    /* ------------------------- modifiers ------------------------- */

    /**
     * Changes the total hours spent on this issue
     *
     * @param amount number of hours (negative to subtract)
     */
    fun addSpent(amount: Double) {
        assert(spent != null) // assert initialized
        local_spent += amount
        assert(spent?.let { it >= 0 } ?: false) // assert non-negative
    }

    /**
     * Changes the estimated hours of this issue
     *
     * @param amount number of hours (negative to subtract)
     */
    fun addEstimated(amount: Double) {
        estimated = estimated?.let {
            // no hours and want to subtract, disable
            if (it == 0.0 && amount < 0) null
            // else change, but don't subtract what can't be subtracted
            else (it + amount).coerceAtLeast(0.0)
        } ?: run {
            // disabled and want to change, set to 0 if it wants to add, keep if not
            if (amount > 0) 0.0 else null
        }
    }

    /**
     * Changes the realization percentage of this issue
     *
     * @param amount percentage (negative to subtract)
     */
    fun addRealization(amount: Int) {
        realization = (realization + amount).coerceIn(0, 100) // don't subtract what can't be subtracted, and don't add what can be added
    }

    /** Sets the realization to the spent/estimated percentage, if loaded */
    fun syncRealization() {
        realization = spent_realization?.coerceIn(0, 100) ?: return
    }

    /* ------------------------- downloading ------------------------- */

    /**
     * Loads spent hours (nothing if already initialized)
     * Long operation
     * returns true if downloaded, false if already made
     *
     * @throws IOException on network error
     */
    @Throws(IOException::class)
    fun downloadExtra(): Boolean {
        remote_spent?.let { return false } // skip initialized

        // load
        val rawIssue = remote.downloadRawIssueDetails(id)
        remote_spent = rawIssue.noNaNDouble("spent_hours")
        journals = rawIssue.optJSONArray("journals")?.run { (0 until length()).map { getJSONObject(it).optString("notes", "") } } ?: emptyList()
        return true
    }

    /** returns this issue's project OT code */
    @Throws(IOException::class)
    fun getProjectOT(): String = remote.getProjectOT(projectId)


    /* ------------------------- uploading ------------------------- */

    /** a list of changes made to this issue, as an object (empty means no changes) */
    internal val changes
        get() = JSONObject().apply {
            // TODO: maybe use a JSON as container so this can be a simple foreach diff
            if (original?.run { noNaNDouble("estimated_hours") != estimated } != false) { // if original is null or estimated is different (true != false)
                // changed hours
                put("estimated_hours", estimated ?: "")
            }
            if (realization != original?.optInt("done_ratio", 0)) {
                // changed ratio
                put("done_ratio", realization)
            }
        }

    /** iff this entry requires upload, false otherwise */
    val requiresUpload get() = !changes.isEmpty
    val changedEstimation get() = changes.has("estimated_hours")
    val changedRealization get() = changes.has("done_ratio")

}

/* ------------------------- utils ------------------------- */

/**
 * Get an optional double associated with a key, or null if there is no such key or if its value is not a number (NaN).
 * If the value is a string, an attempt will be made to evaluate it as a number.
 */
private fun JSONObject.noNaNDouble(key: String) = optDouble(key).takeIf { !it.isNaN() }
